text/0000-destructuring-assignment.md

```diff
@@ -0,0 +1,282 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It could be argued that this feature increases the surface area of the language and thus complexity. However, we feel that by decreasing surprise, it actually makes the language less complex for users.
+- It is possible that these changes could result in some confusing diagnostics. However, we have not found any during testing, and these could in any case be ironed out before stabilisation.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+As we argue above, we believe this change increases the perceived consistency of Rust and improves
+idiomatic code in the presence of mutability, and that the
+implementation is simple and intuitive.
+
+One potential alternative that has been put forth in the past is to allow arbitrary patterns on the left-hand side of an assignment,
+but as discussed above and [extensively in this
+thread](https://github.com/rust-lang/rfcs/issues/372), it is difficult to see how this could work in
+practice (especially with complex left-hand sides that do not simply involve identifiers) and it is not clear that this would have any advantages.
+
+# Prior art
+[prior-art]: #prior-art
+
+The most persuasive prior art is Rust itself, which already permits destructuring
+declarations. Intuitively, a declaration is an assignment that also introduces a new binding.
+Therefore, it seems clear that assignments should act similarly to declarations where possible.
+However, it is also the case that destructuring assignments are present in many languages that permit destructuring
+declarations.
+
+- JavaScript
+[supports destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment).
+- Python [supports destructuring assignment](https://blog.tecladocode.com/destructuring-in-python/).
+- Perl
+[supports destructuring assignment](https://perl6advent.wordpress.com/2017/12/05/day-5-destructure-your-arguments-with-perl-6-signatures/).
+- And so on...
+
+It is a general pattern that languages support destructuring assignment when they support
+destructuring declarations.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- The implementation already supports destructuring of every class of expressions that currently make
+sense in Rust. This feature naturally should be extended to any new class of expressions for which
+it makes sense.
+- It could make sense to permit
+[destructuring compound assignments](#Compound-destructuring-assignment) in the future, though we
+defer this question for later discussions.
+- It could make sense to permit [`ref` and `&`](#Unsupported-patterns) in the future.
+- It [has been suggested](https://github.com/rust-lang/rfcs/issues/372#issuecomment-365606878) that
+mixed declarations and assignments could be permitted, as in the following:
```

Comment:
The inverse also seems reasonable- add syntax for an identifier pattern to assign to an existing variable rather than create a new one. That might largely make destructing assignment redundant, and might be a smaller extension to the language.

Reply:
Could you give an example of what you're imagining? The only way I could see this working is by using a new keyword to specify a destructuring assignment, but functionally this would be exactly the same. Note that the syntax has to be at least somewhat similar to what we have here, because we don't just allow assignment to patterns: we allow assignment to fields, etc. which are not valid patterns.

Reply:
I'm imagining extending the existing pattern syntax something like this:

```rust
// old_ident is already in scope, and is written to here:
let (<magic keyword> old_ident, new_ident) = some_tuple;
```

This approach doesn't need any sort of overlap between the expression and pattern grammars, it just introduces a new kind of identifier pattern to the grammar.

Reply:
This would cover assignment to identifiers, but not any other kind of assignable expression (i.e. lvalue), like paths, function calls returning mutable references, etc. I can add it as an alternative, though.

Reply:
I'm not sure why it wouldn't cover other assignable expressions- `<magic keyword>` can prime the parser to expect any place expression.

Reply:
Ah, in that case, then I agree that this should work; I shall add it as an alternative. However, I don't think this would be a natural extension to the language.
- It requires a new keyword or overloading an existing one.
- It is something that needs to be learnt (whereas I would argue that attempting destructuring assignment with the syntax we propose is something people already try naturally).
- It changes the meaning of `let` (which has previously been associated only with binding new variables).
- To be consistent, we would probably want to allow `let <magic keyword> x = value;`, which introduces another way to simply write `x = value;`.
- It is longer and no more readable than the proposed syntax.

Reply:
I find @varkor's arguments very convincing. And in any case, this is a *future extension*, so I don't think it makes sense to discuss such details in this RFC.

Reply:
I think that's a good summary of the downsides. Just want to make sure it's considered as an alternative, given the relative scopes of the two approaches.

Reply:
This idea had been raised before in [the internals](https://internals.rust-lang.org/t/idea-placepattern/11090?u=crlf0710).

Reply:
In addition to what @varkor wrote above, I would suggest that people could simply pre-declare the variables they would like to declare in a mixed assignment. So instead of

```
// old_ident is already in scope, and is written to here:
let (<magic keyword> old_ident, new_ident) = some_tuple;
```

one would use

```
// old_ident is already in scope
let new_indent;  // new_indent is now also in scope
(old_ident, new_ident) = some_tuple;  // write to both
```